### PR TITLE
Move feat from %sqlrender to %sqlcmd snippets

### LINKED ISSUE #647: move feature from `%sqlrender`  -> `%sqlcmd snippets`
we're going to deprecate `%sqlrender`, we should move the logic to `%sqlcmd snippets`

currently, this is how it works:

```
%%sql --save penguins
select * from penguins.csv limit 10
```

```
%%sql --save subset
select * from penguins where island is not null
```

```
cte = %sqlrender subset
print(cte)
```

prints:

```
WITH penguins AS (
select * from penguins.csv limit 10
)
select * from penguins where island is not null
```

Now, with the change, we should get the same when executing: `%sqlcmd snippets subset`

i.e., 

```
cte = %sqlcmd snippets subset
print(cte)
```

should print:

```
WITH penguins AS (
select * from penguins.csv limit 10
)
select * from penguins where island is not null
```

### PR DESCRIPTION
**Pull request by bbeat2782**
## Describe your changes

Moved `%sqlrender` feature to `%sqlcmd snippets`, added FutureWarning when user call `%sqlrender`, modified API Reference documentation, and modified test functions to test `%sqlcmd snippets {snippet_name}`

## Issue number

Closes #647 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--656.org.readthedocs.build/en/656/

### CODE REVIEW

File: src/sql/sqlcmd.py [hunk 1]
```diff
@@ -58,6 +59,8 @@ def sqlcmd_snippets(others):
         help="Force delete all stored snippets",
         required=False,
     )
+    if len(others) == 1 and others[0] in util.get_all_keys():
```
Comment (edublancas) [inline src/sql/sqlcmd.py:62]: let's show a more informative error. currently:

```
In [7]: %sqlcmd snippets not-a-snippet
UsageError: unrecognized arguments: not-a-snippet
```

it'd be better to show:

```
UsageError: 'not-a-snippet' is not a snippet. Available snippets are A, B, and C
```

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/sql/sqlcmd.py (hunk 1 above)
Comment (bbeat2782) [reply src/sql/sqlcmd.py:62]: > let's show a more informative error. currently:
> 
> ```
> In [7]: %sqlcmd snippets not-a-snippet
> UsageError: unrecognized arguments: not-a-snippet
> ```
> 
> it'd be better to show:
> 
> ```
> UsageError: 'not-a-snippet' is not a snippet. Available snippets are A, B, and C
> ```

Okay. I will work on that after resolving the commit error for this PR.

Comment (edublancas) [reply src/sql/sqlcmd.py:62]: I think your PR is fine, the error I see in the logs is:

```
 FAILED src/tests/test_ggplot.py::test_ggplot_geom_histogram_with[png] - _tkinter.TclError: Can't find a usable init.tcl in the following directories: 
    {C:\hostedtoolcache\windows\Python\3.11.4\x64\tcl\tcl8.6}

C:/hostedtoolcache/windows/Python/3.11.4/x64/tcl/tcl8.6/init.tcl: couldn't read file "C:/hostedtoolcache/windows/Python/3.11.4/x64/tcl/tcl8.6/init.tcl": No error
couldn't read file "C:/hostedtoolcache/windows/Python/3.11.4/x64/tcl/tcl8.6/init.tcl": No error
    while executing
"source C:/hostedtoolcache/windows/Python/3.11.4/x64/tcl/tcl8.6/init.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel #0 [list source $tclfile]"

```

this is an intermittent error, if you push an empty commit to trigger the CI, it should pass:

```
git commit --allow-empty -m ci
```

**Review by yafimvo — COMMENTED** [review]
@bbeat2782 
lgtm. Please rebase.

@edublancas do you think we should remove the `sqlrender` guide from the doc tree?

image

File: src/sql/sqlcmd.py (hunk 1 above)
Comment (yafimvo) [reply src/sql/sqlcmd.py:62]: @bbeat2782 
I re-ran it and it passes

File: doc/api/magic-snippets.md [hunk 2]
```diff
@@ -73,14 +72,22 @@ Returns all the snippets saved in the environment
...
+`{snippet_name}` Return a snippet.
+
 `-d`/`--delete` Delete a snippet.
 
 `-D`/`--delete-force` Force delete a snippet. This may be useful if there are other dependent snippets, and you still need to delete this snippet.
 
 `-A`/`--delete-force-all` Force delete a snippet and all dependent snippets.
 
 ```{code-cell} ipython3
+cte = %sqlcmd snippets gentoo
```
Comment (neelasha23) [inline doc/api/magic-snippets.md:84]: `cte` -> `gentoo_snippet`

File: doc/api/magic-snippets.md [hunk 3]
```diff
@@ -73,14 +72,22 @@ Returns all the snippets saved in the environment
...
+print(cte)
```
Comment (neelasha23) [inline doc/api/magic-snippets.md:85]: This is currently displaying 
```
WITH
SELECT * FROM penguins.csv where species == 'Gentoo'
```

Is it supposed to display `WITH` even when it is not used as subquery as part of another query?

**Comment by neelasha23** [comment]
Added some comments. Few other observations:

1. There are places in the docs where we use `sqlrender`,e.g., [cte](https://jupysql.ploomber.io/en/latest/compose.html#data-visualization). I think we should replace those as well.
2. Please resolve merge conflicts

**Comment by bbeat2782** [comment]
> 1\. There are places in the docs where we use sqlrender,e.g., [cte](https://jupysql.ploomber.io/en/latest/compose.html#data-visualization). I think we should replace those as well.

I was thinking about replacing %sqlrender with %sqlcmd snippets with a separate PR, but if you think it works better with one PR, I will add those changes to this PR.

File: doc/api/magic-snippets.md (hunk 3 above)
Comment (bbeat2782) [reply doc/api/magic-snippets.md:85]: > 1\. Is it supposed to display WITH even when it is not used as subquery as part of another query?

I was following the behavior of `%sqlrender`, but if only showing `WITH` when a snippet is used as subquery is easier to understand on the user side, I will modify the behavior.

> 2\. Also, previously we could sqlrender the final query as well, that uses a snippet as subquery. e.g.
SELECT * from gentoo where bill_length < 50... can you add a similar example ?

For this one, when you mentioned `sqlrender` showing the final query, are you referring to something like the following?
Screen Shot 2023-06-26 at 1 20 06 PM

Comment (edublancas) [reply doc/api/magic-snippets.md:85]: @neelasha23: yes, this is a bug we discovered recently. I reported it on https://github.com/ploomber/jupysql/issues/657, @bbeat2782 will work on it.

@bbeat2782: please work on this first, and then tackle the other one

**Comment by edublancas** [comment]
> I was thinking about replacing %sqlrender with %sqlcmd snippets with a separate PR, but if you think it works better with one PR, I will add those changes to this PR.

yes, please tackle documentation changes here. usually whenever we modify a feature, the PR that changes the feature also updates the docs.

**Review by edublancas — CHANGES_REQUESTED** [review]
please address the observations from @neelasha23

**Comment by edublancas** [comment]
@bbeat2782: we just merged https://github.com/ploomber/jupysql/pull/654 which moved the implementation of `%sqlcmd` to separate files (same logic, just re-organized in different files). this is causing merge conflicts. check if you can fix them. otherwise it might be better to branch from master and apply your changes manually to the new files

**Comment by bbeat2782** [comment]
> Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.

@edublancas it says `src/sql/sqlcmd.py` is the conflicting file but the `Resolve conflicts` button is disabled from my side with a message saying "Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests."

So it seems like I would have to apply the changes I made manually.

**Comment by edublancas** [comment]
> @edublancas it says src/sql/sqlcmd.py is the conflicting file but the Resolve conflicts button is disabled from my side with a message saying "Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests."

I think you'll be able to do it from the CLI, but yeah feel free to do it manually (create a new branch from master and then apply your changes)

**Pull request closed without merging**